### DISCUSSION TITLE
[Feat req]: Support of elasticsearch cross cluster search

### DISCUSSION THREAD

**Discussion by Elufimov** [grafana/grafana#56259] [Observability]
Cross cluster search will be very useful in some cases.
Elasticsearch [documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-cross-cluster-search.html) 
Kibana [documentation](https://www.elastic.co/guide/en/kibana/current/management-cross-cluster-search.html)

**Comment by reilee** [comment]
+1

**Comment by fredericve** [comment]
That would be a very useful feature indeed. We have multiple elasticsearch clusters distributed in different DC's. With this feature we could point to one single elasticsearch cluster which will then be able to search the data of all different cluster.

**Comment by jehutywong** [comment]
+1

**Comment by h-svab** [comment]
+1

**Comment by juan-domenech** [comment]
**Current status:** The validation of a Elasticsearch datasource configured using a Cross Cluster Search pattern in Index Name fails because Elasticsearch does not provide the method _mapping under Cross Cluster Search (used by the datasource validator).
Example: Index Name [*:logstash-]YYYY.MM.DD translates into:
Request: GET /%2A:logstash-2018.07.24/_mapping
Response: {}

**Workaround:** Ignore the error message at the datasources menu and use the newly created datasource. It works ¯\_(ツ)_/¯

**Comment by Elufimov** [comment]
This is really unexpected! Thx for sharing @juan-domenech  👍

**Comment by EugeneKrapivin** [comment]
@juan-domenech It works but some features that use mappings do not work.
as an example: term completion doesn't work, Explore feature doesn't seem to work as well.

I believe there are more "small" features that do not work at all of as expected.

**Reply by petergvizd** [reply]
what version of grafana are you using/did you used? because in version 9.1.5 it is not working at all, it looks to me like Grafana is executing /mapping request before running search query.

**Comment by aocenas** [comment]
Seems like this is a duplicate of https://github.com/grafana/grafana/issues/8170. Please reopen if I am mistaken.

**Comment by gregoryfranklin** [comment]
https://github.com/grafana/grafana/issues/8170 refers to the ability to provide multiple hosts for a single elasticsearch cluster.  It discusses the service discovery mechanism used to select which hosts to connect to.

This is not the same as cross-cluster search.

Cross-cluster search uses a slightly restricted set of API calls allowing one cluster to federate queries across multiple remote clusters.  This feature request would therefore be to make sure that all the API calls made by Grafana to Elasticsearch are compatible with cross cluster search or that Grafana handles the inability to perform certain types of query gracefully.

**Comment by simianhacker** [comment]
@Elfo404 I was just looking into what it will take to get this working correctly. We need to change the `getFields` query to use `_field_caps` instead of `_mapping` API.

**Comment by yeya24** [comment]
This is a very much-needed feature. I think the way mentioned by @simianhacker is a good way to go.

**Comment by gijs-smsc** [comment]
+1

**Comment by philip-smartbit** [comment]
+1

**Comment by leoswing** [comment]
Any progress? What could we do with remote cross cluster search index mappings?

**Comment by leoswing** [comment]
> 

seems not work for me, I use something like `[*:logstash-]*`, but still gets the errors out with `No date field named date found`, and my Grafana version is 8.5.0, ElasticSearch version is 7.10+

**Comment by gadupushailaja** [comment]
Any progress here, I tried [\*:logs-\*] and all other combinations doesn't seem to work, and errors out with "No date field named date found". Is there any other way we can add the cross cluster search index mapping?

**Comment by gabor** [comment]
NOTE: i did a quick look into replacing the usage of the  `_mapping` endpoint with the `_field_caps` endpoint... one problem is that by-default it returns metadata fields too.  since elastic `7.13.0` it adds an is-metadata field to `_field_caps`, but below `7.13.0` i don't see how we could differentiate between metadata and non-metadata fields.

**Comment by JAAnguloCanovas** [comment]
Hi @gabor, any news on this? This is quite an important problem for our organisation, we have hundreds of teams using elasticsearch and they cannot see the Terms of their indexes because our central cluster does not store indexes and reads them from other individual clusters

Regards

**Reply by gabor** [reply]
hi @JAAnguloCanovas , there are no short-term plans for this, but long-term we are planning to implement this. unfortunately i do not have any time estimates.

**Comment by gabor** [comment]
a github issue was created at https://github.com/grafana/grafana/issues/72353 to track this issue now. feel free to comment on that one in the future. thanks!
 
locking the conversation in this one.
